### PR TITLE
SQL: Improve the optimization of null conditionals

### PR DESCRIPTION
**Pull request by costin**
Enhance the existing rules so that
Coalesce(ex) -> ex
NullIf(a, a) -> null
NullIf(null, a) -> null
NullIf(a, null) -> a

### CODE REVIEW

**Review by matriv — APPROVED** [review]
LGTM. Nice! left a question.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/conditional/NullIf.java [hunk 1]
```diff
@@ -40,9 +44,25 @@ public Expression replaceChildren(List<Expression> newChildren) {
...
+        return left;
+    }
+
+    public Expression right() {
+        return right;
+    }
+
+    @Override
+    public boolean foldable() {
+        return left.semanticEquals(right) || super.foldable();
```
[the remark this answers is not in the document]
Comment (costin) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/conditional/NullIf.java:57]: That's covered by the optimizer rule below: https://github.com/elastic/elasticsearch/pull/71192/files#diff-f89f100e88759827c9f9cbceed83a4efb475521a3e42d489584917807df11190R702

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java [hunk 2]
```diff
@@ -694,27 +696,52 @@ protected Expression rule(Expression e) {
...
+                }
+                if (e instanceof ArbitraryConditionalFunction) {
+                    ArbitraryConditionalFunction c = (ArbitraryConditionalFunction) e;
+
+                    // there's no need for a conditional if all the children are the same (this includes the case of just one value)
+                    if (c instanceof Coalesce && children.size() > 0) {
+                        Expression firstChild = children.get(0);
+                        boolean sameChild = true;
+                        for (int i = 1; i < children.size(); i++) {
+                            Expression child =  children.get(i);
```
Comment (bpintea) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:716]: Nit: there's an extra space after equals sign, but the assignment could also be done away with, `child`'s only used once in following expression.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/conditional/NullIf.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (bpintea) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/conditional/NullIf.java:57]: > but then we need to call left.fold() so I'm not sure we should do that.

@matriv fmi, why would that be problematic?, if `.foldable()`?

**Review by bpintea — APPROVED** [review]
LGTM

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/conditional/NullIf.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (matriv) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/conditional/NullIf.java:57]: Conceptually you call `foldable()` beforehand, to check if it's safe to `fold()` something, so it's weird to call `fold()` on a child inside the check of `foldable()`.

**Pull request merged**